### PR TITLE
Stem performance boost

### LINKED ISSUE #7969: Stem is slow and will crash if I try to close the window
**Bug summary**

Stem will crash if you try to close the window while it's processing.  It's also very slow for what I would consider a relatively small number of data points.  I'm using the code from Julia, but I was informed the problem is Matplotlib.  See original [discusion](https://discourse.julialang.org/t/pyplot-crash/1706).

**Code for reproduction**

using PyPlot
stem(randn(32768))

**Actual outcome**

This will work if you wait about half a minute.  If you interfere by trying to close the window, the application crashes (Julia in this case).

**Matplotlib version**

I don't know.  I'm using the most recent version of PyPlot and PyCall in Julia 0.5.  Conda installed Python 2.7.

### PR DESCRIPTION
**Pull request by dstansby**
Instead of adding each line individually, add them as a line collection. This massively improves performance. Fixes #7969 

- Changes what is stored in `StemContainer` from a list of `Line2D` to a `LineCollection` (have added API change to reflect this)
- There's no way of setting the `linemarker` this way

### CODE REVIEW

**Comment by WeatherGod** [comment]
Looks like some extra files got included in this commit?

File: lib/matplotlib/container.py [hunk 1]
```diff
@@ -123,8 +123,16 @@ def __init__(self, lines, has_xerr=False, has_yerr=False, **kwargs):
...
 class StemContainer(Container):
-
     def __init__(self, markerline_stemlines_baseline, **kwargs):
+        '''
+        Parameters
+        ----------
+        markerline_stemlines_baseline : tuple
+            Tuple of ``(markerline, stemlines, baseline)``.
+            ``markerline`` contains the `LineCollection` of the markers,
+            ``stemlines`` is a list of the `Line2D` of each line,
```
Comment (WeatherGod) [inline lib/matplotlib/container.py]: Should this doc entry be updated to say LineCollection? Does it matter? Should we coerce to one type or another?

**Comment by jklymak** [comment]
@dstansby you are failing on `legend_demo`.  Can we merge #9324 first to make sure it has fixed whatever the error is?  I'd hate for you to chase down errors in the argument handling in code that will change soon (I hope).

**Comment by jklymak** [comment]
BTW, not saying it will be fixed - it looks like legend_demo wants a list of lines....

File: lib/matplotlib/legend_handler.py [hunk 2]
```diff
@@ -542,11 +542,8 @@ def create_artists(self, legend, orig_handle,
 
         leg_stemlines = []
         for thisx, thisy in zip(xdata_marker, ydata):
-            l = Line2D([thisx, thisx], [bottom, thisy])
-            leg_stemlines.append(l)
-
-        for lm, m in zip(leg_stemlines, stemlines):
```
Comment (dstansby) [inline lib/matplotlib/legend_handler.py]: As far as I can tell these two lines aren't needed; not 100% sure though...

**Comment by dstansby** [comment]
I've added a legend to the image test just so `stem` legends are covered during testing and not just during the doc build.

File: doc/api/api_changes/2017-10-24-DS.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
+`StemContainer` now stores `LineCollection`
+-------------------------------------------
+
+`StemContainer` objects now store a `LineCollection` object instead of a list
+of `Line2D` objects for stem lines plotted using `ax.stem`.
```
Comment (tacaswell) [inline doc/api/api_changes/2017-10-24-DS.rst]: Can you expand on the consequences of this a bit?  I would have to go look up the code to see how to do a migration.

Adding a note about the speed up would help mollify the angry user who's code we just broke ;)

Comment (dstansby) [reply doc/api/api_changes/2017-10-24-DS.rst]: I've had a go, but not too sure on what to put - if anyone has a better suggestion feel free to push to my branch.

**Comment by jklymak** [comment]
```python 
x = np.linspace(0.1, 2 * np.pi, 100)
fig, ax = plt.subplots()
ax.stem(x, np.cos(x), linefmt='C0-', markerfmt='k+', basefmt='C1-.',
            label='Stem')
ax.stem(x, np.sin(x), linefmt='k-', markerfmt='r+', basefmt='C1-.',
            label='Sine')
ax.legend()
```

Has a slight blemish in the legend in that the `Sine` line has a bue stem instead of a black one...

figure_1

File: lib/matplotlib/legend_handler.py [hunk 4]
```diff
@@ -619,6 +624,13 @@ def create_artists(self, legend, orig_handle,
 
         return artists
 
+    def _copy_collection_props(self, legend_handle, orig_handle):
```
Comment (dstansby) [inline lib/matplotlib/legend_handler.py]: I managed to fix the legend properties by adding this function, which kind of works.

**Comment by dstansby** [comment]
I've made a slightly hacky fix for the legend.

What I *really* want to do is find a way of extracting a property cycle from the `LineCollection` object, and then apply each set of properties to a `Line2D` object; does anyone know how to do this?

File: doc/api/api_changes/2017-10-24-DS.rst [hunk 5]
```diff
@@ -0,0 +1,10 @@
+`StemContainer` now stores `LineCollection`
+-------------------------------------------
+
+`StemContainer` objects now store a `LineCollection` object instead of a list
+of `Line2D` objects for stem lines plotted using `ax.stem`. This gives a very
+large performance boost to displaying and moving `ax.stem` plots.
+
+Line segments can be extracted from the `LineCollection` using
+`LineCollection.get_segements()`. See the `LineCollection` documentation for
```
Comment (anntzer) [inline doc/api/api_changes/2017-10-24-DS.rst]: typo segments

**Comment by anntzer** [comment]
Can you clarify what you need, and why you're not satisfied with your current implementation of the legend handler?

**Comment by dstansby** [comment]
Hmm, I'm not sure anymore, maybe this way is fine.

**Comment by QuLogic** [comment]
So the legend is slightly off in that they elements are drawn in the wrong order; can that be fixed?

This also needs a rebase.

**Comment by jklymak** [comment]
https://github.com/matplotlib/matplotlib/pull/9565#issuecomment-362754879 still stands - the stem and marker are drawn in the wrong zorder in the legend.  Otherwise this looks fine to me.

**Comment by dstansby** [comment]
I've had a bit of a play around, but I can't find a way to set the zorder in a way that works.

**Comment by tacaswell** [comment]
It may be down to the order you construct the compound legend entry.

**Comment by dstansby** [comment]
Thanks @tacaswell - I think this is now working, lets see what CI says.

**Comment by dstansby** [comment]
👍 looks good to go

File: lib/matplotlib/axes/_axes.py [hunk 6]
```diff
@@ -2421,12 +2424,12 @@ def stem(self, *args, linefmt=None, markerfmt=None, basefmt=None,
         markerline, = self.plot(x, y, color=markercolor, linestyle=markerstyle,
                                 marker=markermarker, label="_nolegend_")
 
-        stemlines = []
+        lines = []
         for thisx, thisy in zip(x, y):
-            l, = self.plot([thisx, thisx], [bottom, thisy],
-                           color=linecolor, linestyle=linestyle,
-                           marker=linemarker, label="_nolegend_")
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: What happened to `marker`?

File: lib/matplotlib/container.py [hunk 7]
```diff
@@ -174,10 +174,17 @@ class StemContainer(Container):
 
     baseline : :class:`~matplotlib.lines.Line2D`
         The artist of the horizontal baseline.
-
     """
-
```
Comment (QuLogic) [inline lib/matplotlib/container.py]: Don't need to remove this line.

File: lib/matplotlib/legend_handler.py [hunk 8]
```diff
@@ -603,31 +602,44 @@ def create_artists(self, legend, orig_handle,
...
                               [bottom, bottom])
 
         self.update_prop(leg_baseline, baseline, legend)
 
-        artists = [leg_markerline]
-        artists.extend(leg_stemlines)
+        leg_markerline = Line2D(xdata_marker, ydata[:len(xdata_marker)])
+        self.update_prop(leg_markerline, markerline, legend)
+
+        artists = leg_stemlines
```
Comment (QuLogic) [inline lib/matplotlib/legend_handler.py]: I think @anntzer would say that you can use Python 3 syntax now:
`artist = [*leg_stemlines, leg_baseline, leg_markerline]`

**Comment by NelleV** [comment]
I've asked on gitter, but might as well ask here as well. If I am understanding the pull request properly, it introduces a fairly important backward incompatible change. Has this been discussed at all?

**Comment by WeatherGod** [comment]
It probably should be discussed further. Maybe figure out a migration
strategy? I agree that the performance boost is very compelling and that
this is the much better data object to store the results in, but yes, we
need to consider backwards compatibility.

**Comment by ImportanceOfBeingErnest** [comment]
What about the usual strategy: add a kwarg `return_type`, use `return_type=None` as default, make that one create the new collection-based stem plot. Allow for `return_type="lines"` and let that one produce the old, backward-compatible version of the stem?

One could also make the `"lines"` default for the course of one release and only change it to the new default after another version.

I think the improvement here is really large enough to allow for the API break, as long as a backward-compatible, and easily accessible version still exist. This is because I do not know of any downstream package making use of `stem` and I don't have the impression it is that widely used anyways. Still completely removing the old version from one release to the next is probably a bad move.

**Comment by jklymak** [comment]
I don't recall the exact conversation, but I think it was considered worth breaking the backwards compatibility.  

I guess what would help in the API change note is a quick summary of before and after of how to do something like change the line width...  I can't imagine its too painful - i.e. old `[l.set_linewidth(2) for l in lines]` versus new `linecol.set_linewidth(2)`.

**Comment by NelleV** [comment]
I think we should consider making the transition more smoothly. Maybe an easy way to provide a two cycle warning is to duplicate the code into a new function, provide the fast version under a new name, and do a two cycle deprecation of this one? I don't know where the name `stem` comes from, and how easy it would be to come up with a new meaningful name. I also don't know whether we could come up with a better way to ensure a two release cycle warning of API change (I didn't give it much thought and just wrote down the first solution I came up with)

**Comment by WeatherGod** [comment]
eh, not exactly. pcolor and pcolormesh are two different things and one was
not intended to replace the other. pcolormesh is faster for a subset of
things, so we couldn't get rid of something that works all the time. Now,
pcolorfast() was a bad design choice because it was trying to figure out
which one to use in what situation.

What we have here is something that will always be faster and more
efficient and we want to deprecate the old approach. We could call this
stem2(), or stemfast() perhaps. And then go a few cycles and get rid of
stem(). A few cycles later, we can take back stem().

**Comment by NelleV** [comment]
The solution I proposed is also the first thing that came to my mind. There may be a better way to ensure backward compatibility for a few release cycles. If LineCollections can be easily (or not) converted on the fly to list of Line2D, then we may be able to exploit that: a solution would then be to use another name than `stemlines` to store the LineCollections (let's call than `stemlines2` for the sake of this explanation). When someone attempts to access `stemlines,` convert the LineCollections to a list of Line2D and raise the deprecation warning. The two attributes `stemlines` and `stemlines2` would be incompatible with one another, but this would give time for people to update their code. In the case where someone would use the `stemlines` attribute, it would also cause this to be significantly slower, but I don't think this matters.

**Comment by dstansby** [comment]
I don't have any bandwidth to work out how to do the deprecation properly, so anyone feel free to push to my branch (please don't rewrite what's already there though)

**Comment by tacaswell** [comment]
Punting to 3.1 to give time to sort out this discussion / put in a more gracefully change strategy.

**Comment by dstansby** [comment]
Had a little think about this; as I've said in the API note,

```
Line segments can be extracted from the `LineCollection` using `LineCollection.get_segements()`.
```

Which is me trying to say that this is how you get the old return type from the new return type, though re-reading it it's not that clear so I will re-word it.

I think @ImportanceOfBeingErnest has the best suggestion, add a kwarg that changes the return type, set as default to the current return type, and warn that the return type will change in 2 release cycles.

**Comment by NelleV** [comment]
The backward compatible kwarg option would only work if by default, stem was not using the optimized version of the code.

I'm not so concerned about providing the user with an option to switch back to the old API, but about giving proper warning to the user on the API breakage.

**Comment by jklymak** [comment]
This PR is still lingering - I don't really know how we give warning that we will deprecate the returned type of a method.  Don't we just change it and in the API note give the way to get the old API back?

**Comment by NelleV** [comment]
@jklymak I don't think this solution has any advantages. If you are going to break someone's code by default and they have to change it anyways to have their code run, you might as well force them to use the new API.
The proper way is to deprecate the old method. There are several ways to do this, and it is just a question of taking the time to do it.

When your library is used by hundreds of thousands of users, you need to be careful about these small modifications that seem minor because you are a core developer and following closely the development of Matplotlib. As a reminder, we broke people's code several time by removing unused module imports in our submodules: while I think we made the right move in removing those unused import, it also underlies the importance of backward compatibility and how small changes affect our users.

The minimal amount of backward compatibility we should aim for is to be able to run the tests and build the document from two version ago with the current version without the tests crashing (understand error, not failures).

**Comment by jklymak** [comment]
I've put on the agenda for Monday's call... I'm not advocating for what @dstansby suggested - I'm advocating we just break the API.  If there is a good way to do that smoothly, I'm all for that.

**Comment by anntzer** [comment]
> I don't think this solution has any advantages. If you are going to break someone's code by default and they have to change it anyways to have their code run, you might as well force them to use the new API.

I don't agree with this point.
- If they were just using `stem()` to plot, well, a stem plot and not do any customization on the artist, then changing the return type doesn't change anything for them.  I'd bet this covers >95% of the use cases.  (Of course, 5% of hundreds of thousands of users is still a lot.)
- If they did, then they'll get a crash (because the returned type is quite different) (unless they're exactly using the artist API that both the old and new return type share, in which case things should *hopefully* not change, but if the APIs are not consistent that should be seen as an opportunity to rationalize our API).  But then they can just replace `stem(...)` by `stem(..., use_old_api=True)` as a quickfix (until we kill the old return type), rather than figuring out how to replace the old API by the new one.

In other words, it gives the users a simple way to temporarily use the old API.  You can even make this an rcparam if you want...

> The proper way is to deprecate the old method. There are several ways to do this, and it is just a question of taking the time to do it.

You proposed https://github.com/matplotlib/matplotlib/pull/9565#issuecomment-404976237, which is essentially (sorry if I misunderstood the proposal?) saying provide `stem_v2` (name up to bikeshedding) with the new semantics and deprecate `stem`; I guess we'd later rename `stem_v2` to `stem`?  Frankly I don't like this approach at all; if you don't ever rename the function to the old name you end up with silly names such as [EnumCalendarInfoExEx](https://docs.microsoft.com/en-us/windows/desktop/api/winnls/nf-winnls-enumcalendarinfoexex) (and stem_v123 isn't really helpful to the code reader either) and if you do that makes the diligent programmer (the one who actually follows each upgrades and cleans up after each deprecation warning) do two changes (stem->stem_v2->stem) instead of a single one (take care of the API change).

**Comment by WeatherGod** [comment]
We could always name the new function `stemplot()`. That won't clutter up
the API with silly looking function calls.

We have been bitten several times before with attempts to change the return
dtype. There are several other libraries that are built on top of us that
provide extra introspective plotting features that depend heavily on the
return type of our functions.

**Comment by anntzer** [comment]
Having subtly different `stem()` and `stemplot()` functions that do nearly the same thing but not really is something that I would definitely consider a poor API...

**Comment by ImportanceOfBeingErnest** [comment]
We have kind of a similar thing going on in the current versions. 

If you try to create an axes at the same subplot position as an existing axes, it will fall back to this existing axes.

```
fig = plt.figure()
ax1 = fig.add_subplot(111)
ax2 = fig.add_subplot(111)
print(ax1 == ax2)  # prints True
```

There is a warning associated with this though: 
```
MatplotlibDeprecationWarning:
Adding an axes using the same arguments as a previous axes currently reuses the earlier instance. 
In a future version, a new instance will always be created and returned.  Meanwhile, this warning
can be suppressed, and the future behavior ensured, by passing a unique label to each axes instance.
```

Which means if you want the (expected and useful) behaviour of creating a new axes at the same position, you need to add a label (or in fact any other unique argument) to the creating function.

```
fig = plt.figure()
ax1 = fig.add_subplot(111)
ax2 = fig.add_subplot(111, label="useless label to get a new axes instance")
print(ax1 == ax2)  # prints False
```

So here the old behaviour is kept, but a strategy of using the new behaviour is presented, i.e. by using a `kwarg`. To apply this concept here (and in contrast to my [previous API breaking proposal](https://github.com/matplotlib/matplotlib/pull/9565#issuecomment-404970373)) one could add `return_type` argument as follows:

* `return_type="lines"`: Old behaviour (lines), no warning.
* `return_type="collection"`: New behaviour (collection), no warning.
* `return_type=None`: The default for now. Sets to "lines".  Raises a warning, like "In a future version the return type will change and stem will return a `LineCollection`. To suppress this warning explicitely state the `return_type`".

**Comment by anntzer** [comment]
I'd much rather do what @ImportanceOfBeingErnest proposed just above.

**Comment by timhoffm** [comment]
I'm also -1 on creating a new function with a modified name and deprecating `stem()` just because we want a different return value.

Also, changing the return value without a prior warning is a massive break.

The `return_type` variant seems the most reasonable, even though that means to have both code paths around for some time.

**Comment by QuLogic** [comment]
Sorry to bikeshed a bit, but if it's going to change _only_ the return value, then it can be called `return_type`. But if it is also changing what gets added to the plot, then it should be named something else like `plotting_method` (or something briefer but as clear). I am not clear on whether the suggested way forward will be the former or the latter.

**Comment by ImportanceOfBeingErnest** [comment]
Well, of course it changes what's in the plot. Two important consquences of this:

* the name of a possible kwarg should be different, and also of course one can think about not using strings but booleans for better accessibility/memorizability like `draw_lines=False` or so. I think all this can be discussed once a general solution is agreed upon. 
* More importantly, just deprecating the return type will not work; the API change in that sense is that the function draws something completely different (but visually identical) to the axes. This makes @NelleV's [suggestion](https://github.com/matplotlib/matplotlib/pull/9565#issuecomment-404994438) quite impractical, because users could equally try to access the stem elements via `ax.get_lines()`.

Also replicating a recent comment in chat here for completeness:   
I did try to search github for `".stem("` to find out how many people might be affected by this. But the search ignores the `.` and the `(` so you end up with massive amounts of results from people visualizing their stem cell research or their scanning transmission electron microscope images. Is there a way to search for literal dots . with GitHub?

**Comment by dopplershift** [comment]
We can't just make this change with no deprecation period. Full stop. This is not fixing a bug, or making a minor improvement and changes how the end image appears somewhat. This is a significant change to how the library is operating in regards to how the call to `stem()` operates, and is readily observed through the return type and public API, like `get_lines()`.

The performance implications are worth making the change, certainly. But, IMO, we don't get to just throw up our hands and say "Doing a deprecation period is too hard". It's possible that code that did all the right things (in the standard code path) with 3.0 will no longer work with 3.1--and will end up in a traceback. As a concrete example, this would completely obliterate code that was making an animated stem plot. That's user-hostile behavior, IMO. If somebody made a similar change to `barbs` and broke MetPy, I'd be seriously pissed. (I may be a tad sensitive as I clean up the breakage from matplotlib 3.0.)

What exactly is wrong with:
```python
def stem(self, *args, **kwargs, use_new_behavior=None):
    if use_new_behavior is None:
        warnings.warn("we're changing this")
        use_new_behavior = False
    if use_new_behavior:
        self._new_stem(*args, **kwargs)
    else:
        self._old_stem(*args, **kwargs)
```
(naming stuff aside) ?

**Comment by anntzer** [comment]
@dopplershift That's literally what was proposed in https://github.com/matplotlib/matplotlib/pull/9565#issuecomment-404970373, and that was deemed no-good-enough in https://github.com/matplotlib/matplotlib/pull/9565#issuecomment-408165490, at which point I did "throw up our hands and said "Doing a deprecation period is too hard"."  But I admittedly misread both the comment and its implication that https://github.com/matplotlib/matplotlib/pull/9565#issuecomment-425918716 would be OK (sorry about that).

---

Not to derail this discussion, but the recent `hold` debacle (https://github.com/matplotlib/matplotlib/pull/12274, both re: cartopy and scipy) did make me wonder what's the point of spending so much energy on this, if the process flow is "we put in the deprecation warning, we remove the feature two minor releases later, downstream libs complain, we put it back, and remove it again later"... looks like we should just shunt the first two steps :)

**Comment by dstansby** [comment]
Since the conversation here is getting a bit ridiculous, I've opened a new PR at #12380

**Pull request closed without merging**